File: clients/terminal/internal/publictext/public_text.go
// Package publictext owns the renderer-safe transform for untrusted text.
package publictext

import (
	"strings"
	"unicode/utf8"
)

const upperHex = "0123456789ABCDEF"

// Transform replaces terminal control code points with inert visible escapes.
// Horizontal tab and line feed are the only admitted C0 controls.
func Transform(value string) string {
	var result strings.Builder
	result.Grow(len(value))
	for len(value) != 0 {
		character, width := utf8.DecodeRuneInString(value)
		if character == utf8.RuneError && width == 1 {
			// Go strings may contain arbitrary bytes.  An invalid UTF-8 byte must
			// never be normalized to U+FFFD because that would hide a C1 control
			// byte from this security boundary.
			result.WriteString(`\x`)
			result.WriteByte(upperHex[value[0]>>4])
			result.WriteByte(upperHex[value[0]&0xF])
			value = value[1:]
			continue
		}
		value = value[width:]
		codePoint := uint32(character)
		if character == '\t' || character == '\n' {
			result.WriteRune(character)
			continue
		}
		if codePoint < 0x20 || (codePoint >= 0x7F && codePoint <= 0x9F) {
			result.WriteString(`\x`)
			result.WriteByte(upperHex[(codePoint>>4)&0xF])
			result.WriteByte(upperHex[codePoint&0xF])
			continue
		}
		result.WriteRune(character)
	}
	return result.String()
}

// IsSafe reports whether Transform would leave the value unchanged.
func IsSafe(value string) bool { return Transform(value) == value }

// Bounded transforms and then truncates without splitting a rune.
func Bounded(value string, maximumRunes, maximumBytes int) string {
	if maximumRunes < 1 || maximumBytes < 1 {
		return ""
	}
	safe := Transform(value)
	var result strings.Builder
	count := 0
	for _, character := range safe {
		if count >= maximumRunes {
			break
		}
		encoded := string(character)
		if result.Len()+len(encoded) > maximumBytes {
			break
		}
		result.WriteString(encoded)
		count++
	}
	return result.String()
}

File: src/pulsara_agent/runtime/terminal_presentation/public_text.py
"""The sole transform from untrusted text to terminal-safe public text."""

from __future__ import annotations


def terminal_safe_public_text(value: str) -> str:
    """Render terminal control code points as inert, visible ASCII escapes."""

    if not isinstance(value, str):
        raise TypeError("terminal public text must be a string")
    parts: list[str] = []
    for character in value:
        code_point = ord(character)
        if character in {"\t", "\n"}:
            parts.append(character)
        elif code_point < 0x20 or 0x7F <= code_point <= 0x9F:
            parts.append(f"\\x{code_point:02X}")
        else:
            parts.append(character)
    return "".join(parts)


def bounded_terminal_safe_public_text(
    value: str,
    *,
    maximum_code_points: int,
    maximum_utf8_bytes: int,
) -> str:
    """Apply the transform, then truncate without splitting a code point."""

    if maximum_code_points < 1 or maximum_utf8_bytes < 1:
        raise ValueError("terminal public text bounds must be positive")
    safe = terminal_safe_public_text(value)
    result: list[str] = []
    encoded_bytes = 0
    for character in safe:
        if len(result) >= maximum_code_points:
            break
        size = len(character.encode("utf-8"))
        if encoded_bytes + size > maximum_utf8_bytes:
            break
        result.append(character)
        encoded_bytes += size
    return "".join(result)


__all__ = ["bounded_terminal_safe_public_text", "terminal_safe_public_text"]
